# Customer & Campaign Intelligence — Feature Engineering

**Proyecto:** Segmentación de clientes y análisis de respuesta a campañas de marketing.

**Objetivo de este notebook:** crear variables derivadas a partir del dataset limpio
(`02_data_cleaning.ipynb`), orientadas a las preguntas de negocio del proyecto:
valor del cliente, comportamiento de compra y respuesta a campañas.

**Entrada:** `data/processed/marketing_campaign_clean.csv`
**Salida:** `data/processed/marketing_campaign_features.csv`

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("../data/processed/marketing_campaign_clean.csv")

In [2]:
print(df.shape)
df


(2236, 27)


,ID,Year_Birth,Education,Marital_Status,Income,Kidhome,Teenhome,Dt_Customer,Recency,MntWines,...,NumCatalogPurchases,NumStorePurchases,NumWebVisitsMonth,AcceptedCmp3,AcceptedCmp4,AcceptedCmp5,AcceptedCmp1,AcceptedCmp2,Complain,Response
0,5524,1957,Graduation,Single,58138.0,0,0,2012-09-04,58,635,...,10,4,7,0,0,0,0,0,0,1
1,2174,1954,Graduation,Single,46344.0,1,1,2014-03-08,38,11,...,1,2,5,0,0,0,0,0,0,0
2,4141,1965,Graduation,Together,71613.0,0,0,2013-08-21,26,426,...,2,10,4,0,0,0,0,0,0,0
3,6182,1984,Graduation,Together,26646.0,1,0,2014-02-10,26,11,...,0,4,6,0,0,0,0,0,0,0
4,5324,1981,PhD,Married,58293.0,1,0,2014-01-19,94,173,...,3,6,5,0,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2231,10870,1967,Graduation,Married,61223.0,0,1,2013-06-13,46,709,...,3,4,5,0,0,0,0,0,0,0
2232,4001,1946,PhD,Together,64014.0,2,1,2014-06-10,56,406,...,2,5,7,0,0,0,1,0,0,0
2233,7270,1981,Graduation,Divorced,56981.0,0,0,2014-01-25,91,908,...,3,13,6,0,1,0,0,0,0,0
2234,8235,1956,Master,Together,69245.0,0,1,2014-01-24,8,428,...,5,10,3,0,0,0,0,0,0,0


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2236 entries, 0 to 2235
Data columns (total 27 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   ID                   2236 non-null   int64  
 1   Year_Birth           2236 non-null   int64  
 2   Education            2236 non-null   object 
 3   Marital_Status       2236 non-null   object 
 4   Income               2236 non-null   float64
 5   Kidhome              2236 non-null   int64  
 6   Teenhome             2236 non-null   int64  
 7   Dt_Customer          2236 non-null   object 
 8   Recency              2236 non-null   int64  
 9   MntWines             2236 non-null   int64  
 10  MntFruits            2236 non-null   int64  
 11  MntMeatProducts      2236 non-null   int64  
 12  MntFishProducts      2236 non-null   int64  
 13  MntSweetProducts     2236 non-null   int64  
 14  MntGoldProds         2236 non-null   int64  
 15  NumDealsPurchases    2236 non-null   i

In [4]:
df["Dt_Customer"] = pd.to_datetime(df["Dt_Customer"])

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2236 entries, 0 to 2235
Data columns (total 27 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   ID                   2236 non-null   int64         
 1   Year_Birth           2236 non-null   int64         
 2   Education            2236 non-null   object        
 3   Marital_Status       2236 non-null   object        
 4   Income               2236 non-null   float64       
 5   Kidhome              2236 non-null   int64         
 6   Teenhome             2236 non-null   int64         
 7   Dt_Customer          2236 non-null   datetime64[ns]
 8   Recency              2236 non-null   int64         
 9   MntWines             2236 non-null   int64         
 10  MntFruits            2236 non-null   int64         
 11  MntMeatProducts      2236 non-null   int64         
 12  MntFishProducts      2236 non-null   int64         
 13  MntSweetProducts     2236 non-nul

In [6]:
df["Dt_Customer"].dt.year.max()

np.int32(2014)

In [7]:
df["Edad"] = 2014 - df["Year_Birth"]

In [8]:
df[["Year_Birth", "Edad"]].head()

,Year_Birth,Edad
0,1957,57
1,1954,60
2,1965,49
3,1984,30
4,1981,33


In [9]:
df[["MntWines", "MntFruits", "MntMeatProducts", "MntFishProducts", "MntSweetProducts", "MntGoldProds"]].head()

,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,MntGoldProds
0,635,88,546,172,88,88
1,11,1,6,2,1,6
2,426,49,127,111,21,42
3,11,4,20,10,3,5
4,173,43,118,46,27,15


In [10]:
df["Gasto_Total"] = df["MntWines"] + df["MntFruits"] + df["MntMeatProducts"] + df["MntFishProducts"] + df["MntSweetProducts"] + df["MntGoldProds"]

In [11]:
df[["MntWines", "MntFruits", "MntMeatProducts", "MntFishProducts", "MntSweetProducts", "MntGoldProds", "Gasto_Total"]].head()

,MntWines,MntFruits,MntMeatProducts,MntFishProducts,MntSweetProducts,MntGoldProds,Gasto_Total
0,635,88,546,172,88,88,1617
1,11,1,6,2,1,6,27
2,426,49,127,111,21,42,776
3,11,4,20,10,3,5,53
4,173,43,118,46,27,15,422


In [12]:
df["Compras_Totales"] = df["NumWebPurchases"] + df["NumCatalogPurchases"] + df["NumStorePurchases"]

In [13]:
df[["NumWebPurchases", "NumCatalogPurchases", "NumStorePurchases", "Compras_Totales"]].head()

,NumWebPurchases,NumCatalogPurchases,NumStorePurchases,Compras_Totales
0,8,10,4,22
1,1,1,2,4
2,8,2,10,20
3,2,0,4,6
4,5,3,6,14


In [14]:
df["Campañas_Aceptadas"] = df["AcceptedCmp1"] + df["AcceptedCmp2"] + df["AcceptedCmp3"] + df["AcceptedCmp4"] + df["AcceptedCmp5"]

In [15]:
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Response', 'Edad', 'Gasto_Total',
       'Compras_Totales', 'Campañas_Aceptadas'],
      dtype='object')

In [16]:
df["Dependientes"] = df["Kidhome"] + df["Teenhome"]


In [17]:
df[["Kidhome","Teenhome"]].head()

,Kidhome,Teenhome
0,0,0
1,1,1
2,0,0
3,1,0
4,1,0


In [18]:
fecha_referencia = df["Dt_Customer"].max()
print(fecha_referencia)

2014-06-29 00:00:00


In [19]:
df["Antiguedad_Cliente"] = fecha_referencia - df["Dt_Customer"]

In [20]:
df[["Dt_Customer", "Antiguedad_Cliente"]].head()

,Dt_Customer,Antiguedad_Cliente
0,2012-09-04,663 days
1,2014-03-08,113 days
2,2013-08-21,312 days
3,2014-02-10,139 days
4,2014-01-19,161 days


In [21]:
df["Antiguedad_Cliente"] = df["Antiguedad_Cliente"].dt.days

In [22]:
df[["Dt_Customer", "Antiguedad_Cliente"]].head()

,Dt_Customer,Antiguedad_Cliente
0,2012-09-04,663
1,2014-03-08,113
2,2013-08-21,312
3,2014-02-10,139
4,2014-01-19,161


In [23]:
print(df.shape)
df.columns

(2236, 33)


Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Response', 'Edad', 'Gasto_Total',
       'Compras_Totales', 'Campañas_Aceptadas', 'Dependientes',
       'Antiguedad_Cliente'],
      dtype='object')

In [24]:
df["Gasto_Total"].quantile([0.33, 0.66])

0.33    110.65
0.66    813.20
Name: Gasto_Total, dtype: float64

In [25]:
df["Segmento_Valor"] = pd.cut(
    df["Gasto_Total"],
    bins=[-1, 110.65, 813.20, df["Gasto_Total"].max()],
    labels=["Bajo", "Medio", "Alto"]
)

In [26]:
df["Segmento_Valor"].value_counts()

Segmento_Valor
Alto     760
Bajo     738
Medio    738
Name: count, dtype: int64

In [27]:
pd.crosstab(df["Segmento_Valor"], df["Response"])

Response,0,1
Segmento_Valor,,
Bajo,682,56
Medio,647,91
Alto,573,187


In [28]:
df.columns

Index(['ID', 'Year_Birth', 'Education', 'Marital_Status', 'Income', 'Kidhome',
       'Teenhome', 'Dt_Customer', 'Recency', 'MntWines', 'MntFruits',
       'MntMeatProducts', 'MntFishProducts', 'MntSweetProducts',
       'MntGoldProds', 'NumDealsPurchases', 'NumWebPurchases',
       'NumCatalogPurchases', 'NumStorePurchases', 'NumWebVisitsMonth',
       'AcceptedCmp3', 'AcceptedCmp4', 'AcceptedCmp5', 'AcceptedCmp1',
       'AcceptedCmp2', 'Complain', 'Response', 'Edad', 'Gasto_Total',
       'Compras_Totales', 'Campañas_Aceptadas', 'Dependientes',
       'Antiguedad_Cliente', 'Segmento_Valor'],
      dtype='object')

In [29]:
df.to_csv("../data/processed/marketing_campaign_features.csv", index=False)

## Feature Engineering — Resumen

A partir del dataset limpio (2,236 filas × 27 columnas), se crearon las 
siguientes variables:

1. **Edad:** 2014 (año más reciente en `Dt_Customer`) − `Year_Birth`.
2. **Gasto_Total:** suma del gasto en las 6 categorías de producto 
   (`MntWines`, `MntFruits`, `MntMeatProducts`, `MntFishProducts`, 
   `MntSweetProducts`, `MntGoldProds`).
3. **Compras_Totales:** suma de compras por los 3 canales 
   (`NumWebPurchases`, `NumCatalogPurchases`, `NumStorePurchases`).
4. **Campañas_Aceptadas:** suma de las 5 campañas anteriores aceptadas 
   (`AcceptedCmp1` a `AcceptedCmp5`).
5. **Dependientes:** suma de `Kidhome` + `Teenhome`.
6. **Antiguedad_Cliente:** días transcurridos entre `Dt_Customer` y la 
   fecha más reciente del dataset (2014-06-29).
7. **Segmento_Valor:** clasificación en Bajo/Medio/Alto según percentiles 
   (33% y 66%) de `Gasto_Total`.

**Resultado final:** 2,236 filas × 34 columnas, guardado en 
`data/processed/marketing_campaign_features.csv`.